### PR TITLE
Prometheus: Create unit test project

### PR DESCRIPTION
**Pull request by CodeBlanch**
## Changes

Getting started on Prometheus unit tests. This creates a project and adds a simple test.

### CODE REVIEW

**Review by pellared — APPROVED** [review]

File: test/OpenTelemetry.Exporter.Prometheus.Tests/PrometheusExporterExtensionsTests.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+using Xunit;
+
+namespace OpenTelemetry.Exporter.Prometheus.Tests
+{
+    public sealed class PrometheusExporterExtensionsTests : IDisposable
+    {
+        public PrometheusExporterExtensionsTests()
+        {
+            // Fix the times so tests can be deterministic
+            PrometheusMetricBuilder.GetUtcNowDateTimeOffset = () => DateTimeOffset.Parse("9/30/2021 10:30pm +00:00");
```
Comment (reyang) [inline test/OpenTelemetry.Exporter.Prometheus.Tests/PrometheusExporterExtensionsTests.cs:33]: Consider avoiding "Parse" and use numbers+ctor?

From the [official doc](https://docs.microsoft.com/en-us/dotnet/api/system.datetimeoffset.parse?view=net-5.0):

_The input string is parsed by using the formatting information in a DateTimeFormatInfo object that is initialized for the **current culture**._

File: src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusMetricBuilder.cs [hunk 2]
```diff
@@ -28,6 +28,8 @@ internal sealed class PrometheusMetricBuilder
     {
         public const string ContentType = "text/plain; version = 0.0.4";
 
+        internal static Func<DateTimeOffset> GetUtcNowDateTimeOffset = () => DateTimeOffset.UtcNow;
```
Comment (reyang) [inline src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusMetricBuilder.cs:31]: Do we really need/want this to be static?

File: test/OpenTelemetry.Exporter.Prometheus.Tests/PrometheusExporterExtensionsTests.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            using var meter = new Meter("TestMeter", "0.0.1");
+
+            var counter = meter.CreateCounter<int>("counter");
+
+            counter.Add(100, tags);
+
+            var testCompleted = false;
+
+            // Invokes the TestExporter which will invoke RunTest
+            metricReader.Collect();
```
Comment (reyang) [inline test/OpenTelemetry.Exporter.Prometheus.Tests/PrometheusExporterExtensionsTests.cs:66]: Consider this:
1. Change Ln54 to `.AddReader(new BaseExportingMetricReader(new TestExporter<Metric>(RunTest)))` (no need to have the `metricReader` variable.
2. Change this line to `provider.ForceFlush(someTimeout)`.

File: src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusMetricBuilder.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusMetricBuilder.cs:31]: Makes the tests much easier to write. What's your issue with it? Note the value isn't static, it is wrapping `DateTimeOffset.UtcNow` access into a delegate which the tests can swap out to make things deterministic.

Comment (reyang) [reply src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusMetricBuilder.cs:31]: My worry is that if folks use this in multiple test cases and they run the tests simultaneously, it'll cause race condition (with different unit tests expecting different angel clocks)

File: test/OpenTelemetry.Exporter.Prometheus.Tests/PrometheusExporterExtensionsTests.cs (hunk 1 above)
Comment (CodeBlanch) [reply test/OpenTelemetry.Exporter.Prometheus.Tests/PrometheusExporterExtensionsTests.cs]: Updated

File: test/OpenTelemetry.Exporter.Prometheus.Tests/PrometheusExporterExtensionsTests.cs (hunk 3 above)
Comment (CodeBlanch) [reply test/OpenTelemetry.Exporter.Prometheus.Tests/PrometheusExporterExtensionsTests.cs]: Updated

File: src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusMetricBuilder.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusMetricBuilder.cs]: I was thinking we would just control the parallelism of the WriteMetricsCollection tests, but it was pretty easy to make it instance-based so I changed it.

File: test/OpenTelemetry.Exporter.Prometheus.Tests/PrometheusExporterExtensionsTests.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        {
+            var tags = new KeyValuePair<string, object>[]
+            {
+                new KeyValuePair<string, object>("key1", "value1"),
+                new KeyValuePair<string, object>("key2", "value2"),
+            };
+
+            using var provider = Sdk.CreateMeterProviderBuilder()
+                .AddSource("TestMeter")
+                .AddReader(new BaseExportingMetricReader(new TestExporter<Metric>(RunTest)))
```
Comment (cijothomas) [inline test/OpenTelemetry.Exporter.Prometheus.Tests/PrometheusExporterExtensionsTests.cs:41]: Do we want to add a functional test, where Prometheus spins up a server/middleware, and the test make a webrequest to it, and validate the response contents?

**Review by reyang — APPROVED** [review]

File: test/OpenTelemetry.Exporter.Prometheus.Tests/PrometheusExporterExtensionsTests.cs (hunk 4 above)
Comment (CodeBlanch) [reply test/OpenTelemetry.Exporter.Prometheus.Tests/PrometheusExporterExtensionsTests.cs:41]: Definitely. My plan is to keep adding tests over multiple PRs until we reach some level of acceptable coverage.

**Comment by cijothomas** [comment]
Towards item 3 from https://github.com/open-telemetry/opentelemetry-dotnet/issues/2400

File: test/OpenTelemetry.Exporter.Prometheus.Tests/PrometheusExporterExtensionsTests.cs (hunk 4 above)
Comment (cijothomas) [reply test/OpenTelemetry.Exporter.Prometheus.Tests/PrometheusExporterExtensionsTests.cs:41]: https://github.com/open-telemetry/opentelemetry-dotnet/pull/2030/files#diff-8c9aaed37ddc4021afc4e60bcdb8c381ebcd586c53808ea0634f2946c3173218 I had some basic one for that in the "old" metrics world. Not likely to be very useful, but just sharing.

**Review by cijothomas — APPROVED** [review]

**Pull request merged**